feature: Add "!cmd" user shell execution

abhishek-oai opened the pull request:
feature: Add "!cmd" user shell execution

This change lets users run local shell commands directly from the TUI by prefixing their input with ! (e.g. !ls). Output is truncated to keep the exec cell usable, and Ctrl-C cleanly
  interrupts long-running commands (e.g. !sleep 10000).

**Summary of changes**

  - Route Op::RunUserShellCommand through a dedicated UserShellCommandTask (core/src/tasks/user_shell.rs), keeping the task logic out of codex.rs.
  - Reuse the existing tool router: the task constructs a ToolCall for the local_shell tool and relies on ShellHandler, so no manual MCP tool lookup is required.
  - Emit exec lifecycle events (ExecCommandBegin/ExecCommandEnd) so the TUI can show command metadata, live output, and exit status.

**End-to-end flow**

  **TUI handling**

  1. ChatWidget::submit_user_message (TUI) intercepts messages starting with !.
  2. Non-empty commands dispatch Op::RunUserShellCommand { command }; empty commands surface a help hint.
  3. No UserInput items are created, so nothing is enqueued for the model.

  **Core submission loop**
  4. The submission loop routes the op to handlers::run_user_shell_command (core/src/codex.rs).
  5. A fresh TurnContext is created and Session::spawn_user_shell_command enqueues UserShellCommandTask.

  **Task execution**
  6. UserShellCommandTask::run emits TaskStartedEvent, formats the command, and prepares a ToolCall targeting local_shell.
  7. ToolCallRuntime::handle_tool_call dispatches to ShellHandler.

  **Shell tool runtime**
  8. ShellHandler::run_exec_like launches the process via the unified exec runtime, honoring sandbox and shell policies, and emits ExecCommandBegin/End.
  9. Stdout/stderr are captured for the UI, but the task does not turn the resulting ToolOutput into a model response.

  **Completion**
  10. After ExecCommandEnd, the task finishes without an assistant message; the session marks it complete and the exec cell displays the final output.

  **Conversation context**

  - The command and its output never enter the conversation history or the model prompt; the flow is local-only.
  - Only exec/task events are emitted for UI rendering.

**Demo video**

https://github.com/user-attachments/assets/fcd114b0-4304-4448-a367-a04c43e0b996

easong-openai commented:
We need to make sure cancellation works, particularly ctr-c. Right now it doesn't actually cancel the command.

If possible, we should show the command we just ran + any output and delineate it as a user command instead of using the shortening UI we do for the model.

Otherwise nice work!

easong-openai commented:
This also doesn't support windows, which isn't necessarily a blocker but we might need to look at.

abhishek-oai commented:
I have read the CLA Document and I hereby sign the CLA

abhishek-oai commented:
> We need to make sure cancellation works, particularly ctr-c. Right now it doesn't actually cancel the command.
> 
> If possible, we should show the command we just ran + any output and delineate it as a user command instead of using the shortening UI we do for the model.
> 
> Otherwise nice work!

Alright both those things should work.

Also, the Tui still remains like the "frontend" and forwards the request to the "core". This is in line with the existing archiecture.

abhishek-oai commented on codex-rs/tui/src/chatwidget.rs:
```diff
@@ -548,7 +548,62 @@ impl ChatWidget<'_> {
+                    }
+                } else {
+                    self.submit_user_message(text.into());
+                }
```
Provided in new implementation.

abhishek-oai commented on codex-rs/tui/src/chatwidget.rs:
```diff
@@ -548,7 +548,62 @@
+                            tx.send(AppEvent::LocalExecEnd {
+                                command,
+                                parsed,
+                                output: CommandOutput { exit_code, stdout, stderr },
```
Done.

tibo-openai commented:
@codex review in depth

easong-openai commented:
Seems worth addressing the codex review notes above ^

abhishek-oai commented on codex-rs/tui/src/history_cell.rs:
```diff
@@ -175,6 +180,23 @@ impl WidgetRef for &ExecCell {
+    for raw in src.lines() {
+        lines.push(ansi_escape_line(raw));
+    }
+    PlainHistoryCell { lines }
```
Fixed

abhishek-oai commented:
@codex review in depth

abhishek-oai commented on codex-rs/core/src/local_exec.rs:
```diff
@@ -0,0 +1,55 @@
+        && let Some(pgid) = guard.take()
+    {
+        unsafe {
+            let _ = libc::kill(-pgid, libc::SIGINT);
```
This is fixed now and there is a stub WindowsImpl which we can use for a future Win implementation (if required)

abhishek-oai commented:
> Seems worth addressing the codex review notes above ^

All fixed!

takatoshi-maeda commented:
Does this pull request have a planned release schedule?

When I was using a different coding agent, I was able to provide the necessary context with the ! command—for example, executing database queries or checking schemas. Supplying context through the ! command also helped the coding agent experiment autonomously.

With Codex, I’ve been following this workflow by copying and pasting from the shell, which requires some additional steps. For that reason, I’m very much looking forward to the release of this feature.

abhishek-oai committed:
feature: Add "!cmd" user shell execution

- This feature lets users execute a command using their shell when they type "!cmd" in the TUI for e.g. "!ls"
- It takes care to bound the output line limit to not flood the TUI
- Being able to signal (Ctrl-C) commands spawned. For e.g. "!sleep 10000" killed via SIGINT (Ctrl-C)
- E2E flow of the feature is detailed below

1. The composer hands the submitted text to `ChatWidget::submit_user_message` (`codex-rs/tui/src/chatwidget.rs`).
2. If the message starts with `!`, the prefix is stripped. A non-empty command triggers `ChatWidget::submit_op(Op::RunUserShellCommand { command })`, while empty input shows a help hint instead of contacting the model.
3. Because no `UserInput` items are emitted, the command never joins the conversation payload that the model sees.

4. The submission loop in `codex-rs/core/src/codex.rs` routes the op to `handlers::run_user_shell_command`.
5. The handler creates a new `TurnContext` (with default per-turn settings) and calls `Session::spawn_user_shell_command`, which enqueues a `UserShellCommandTask` without any model-facing input.

6. `UserShellCommandTask::run` (defined in `codex-rs/core/src/tasks/user_shell.rs`) emits a `TaskStartedEvent`, formats the command using the session’s shell wrapper, and creates a `ToolCall` for the built-in `local_shell` tool.
7. `ToolCallRuntime::handle_tool_call` dispatches this call to the `ShellHandler` implementation in `codex-rs/core/src/tools/handlers/shell.rs`.

8. `ShellHandler::run_exec_like` launches the command through the unified exec runtime, honoring sandbox policy and shell environment settings. It uses `ToolEmitter::shell` to publish `ExecCommandBegin` and `ExecCommandEnd` events so the TUI can render the command, streaming output, exit code, and duration.
9. The command output is captured locally for the UI; the handler returns a `ToolOutput::Function`, but the owning task does not turn that into a follow-up assistant message.

- Neither the command nor its stdout/stderr are appended to the conversation history supplied to the model. The bang command is purely a local exec flow; the model receives no additional context from it.
- The only protocol traffic is the pair of exec events and task lifecycle notifications needed by the UI.

10. After `ExecCommandEnd` fires, `UserShellCommandTask` finishes with `None` as the final agent message. The session marks the task complete, and the UI shows the finished exec cell result.

In short: `!cmd` spins up a local shell execution pipeline, streams the output to the user interface, and keeps the model prompt untouched.

abhishek-oai commented:
@bolinfest Updated the implementation to align with our discussion offline

abhishek-oai commented on codex-rs/core/src/shell.rs:
```diff
@@ -133,6 +133,31 @@ fn strip_bash_lc(command: &Vec<String>) -> Option<String> {
 fn detect_default_user_shell() -> Shell {
+    // In tests, prefer a hermetic shell to avoid sourcing user configs.
+    // Try zsh first, then bash; use /dev/null as rc.
+    #[cfg(test)]
```
Happy to make this pass in CI once the PR is reviewed and the approach is agreed on.

bolinfest replied:
@jif-oai do you have thoughts on how better to do this?

abhishek-oai replied:
Added an "env::var" to signal a test and then create a test shell

bolinfest commented on codex-rs/core/src/codex.rs:
```diff
@@ -1474,6 +1478,101 @@ async fn submission_loop(
+
+                    // Build a shell invocation in the user's default shell.
+                    let shell_invocation = sess_clone
+                        .user_shell
+                        // Why we pass a ["bash", "-lc", <script>] sentinel instead of the raw command:
+                        // - The shell adapter (core/src/shell.rs) first calls `strip_bash_lc`. When it sees this
+                        //   exact shape it extracts <script> and then builds the correct argv for the user shell
+                        //   (e.g., `/bin/zsh -lc "source ~/.zshrc && (<script>)"`).
+                        // - If we pass the whole command as a single string (e.g., ["cat Cargo.toml | wc -l"]) the
+                        //   adapter may quote it when joining/embedding, and shells can treat the entire value as a
+                        //   single program name or a single quoted token.
```
Instead of relying on this quirk of `format_default_shell_invocation()`, we should probably just support this directly as a method of `Shell`?

bolinfest commented on codex-rs/core/src/codex.rs:
```diff
@@ -1474,6 +1478,101 @@ async fn submission_loop(
+                    let params = ExecParams {
+                        command: shell_invocation.clone(),
+                        cwd: turn_context.cwd.clone(),
+                        timeout_ms: None,
```
I guess the user can ctrl-c if the command runs too long? Will that work?

abhishek-oai replied:
Yep, tested as well. This is why I leaned on the existing set_task abstraction.

Here is the exact flow as well as per codex -

- Hitting Ctrl+C triggers the same interrupt path we already use for long model turns. The shortcut is caught in the chat widget (tui/src/chatwidget.rs:776-
  785), and because the bottom pane reports a running task, it immediately submits Op::Interrupt instead of shutting the app down (tui/src/chatwidget.rs:1382-1393). 
  
  - That op winds up in Session::interrupt_task, which aborts the currently registered task (core/src/codex.rs:1017-1024); the user shell command was registered as that task when we spawned it (core/src/codex.rs:1555-1570).

-Aborting the task cancels the run_user_shell_command future. The child process that is executing the user’s shell is owned by that future, and because we spawn it with kill_on_drop(true)
  (core/src/spawn.rs:88-106), dropping the task handle issues a kill to the child. So the command receives a termination signal and stops. As part of the abort flow we also emit a
  TurnAborted event, which clears the “command running” state in the UI (core/src/codex.rs:1159-1173), so everything resets cleanly once the child dies.

abhishek-oai replied:
The video has a demo of this too

bolinfest commented on codex-rs/core/src/codex.rs:
```diff
@@ -1474,6 +1478,101 @@ async fn submission_loop(
+                    // Use a fresh diff tracker (no patch application expected for ! commands).
+                    let mut turn_diff_tracker = TurnDiffTracker::new();
+                    // Initiated by user, not by the model. Hence, we generate a new call_id.
+                    let call_id = format!("call_{}", Uuid::new_v4());
```
Hmm, as far as I know we don't require a certain format for `call_id`, but this is something to look out for.

bolinfest commented on codex-rs/core/src/codex.rs:
```diff
@@ -1474,6 +1478,101 @@ async fn submission_loop(
+                            ExecInvokeArgs {
+                                params,
+                                sandbox_type: SandboxType::None,
+                                sandbox_policy: &turn_context.sandbox_policy,
```
Since this is user-initiated, should it be run without a sandbox?

```suggestion
                                sandbox_policy: SandboxPolicy::DangerFullAccess,
```

bolinfest commented on codex-rs/core/src/codex.rs:
```diff
@@ -1474,6 +1478,101 @@ async fn submission_loop(
                 };
                 sess.send_event(event).await;
             }
+            Op::RunUserShellCommand { command } => {
```
This `submission_loop()` function needs to be refactored, so rather than pile on, can we move the handler for `Op::RunUserShellCommand` into its own function?

bolinfest commented on codex-rs/core/tests/suite/user_shell_cmd.rs:
```diff
@@ -0,0 +1,112 @@
+    let file_name = "hello.txt";
+    let file_path: PathBuf = cwd.path().join(file_name);
+    let contents = "hello from bang test\n";
+    fs::write(&file_path, contents).expect("write temp file");
```
Might as well prefer `tokio::fs` since your test function is `async`.

abhishek-oai replied:
Done

bolinfest commented on codex-rs/tui/src/history_cell.rs:
```diff
@@ -426,17 +427,28 @@ impl ExecCell {
+                } else {
+                    TOOL_CALL_MAX_LINES
+                };
+                let out = output_lines(Some(output), false, false, false, line_limit)
```
The multiple booleans make this hard to follow, so I just put up https://github.com/openai/codex/pull/3591.

abhishek-oai replied:
Rebased on top

bolinfest reviewed (changes requested):
Can you please add a video to the PR body?

abhishek-oai commented:
> Does this pull request have a planned release schedule?
> 
> When I was using a different coding agent, I was able to provide the necessary context with the ! command—for example, executing database queries or checking schemas. Supplying context through the ! command also helped the coding agent experiment autonomously.
> 
> With Codex, I’ve been following this workflow by copying and pasting from the shell, which requires some additional steps. For that reason, I’m very much looking forward to the release of this feature.

Actively working to land this.

takatoshi-maeda commented:
Thank you, I really appreciate it. I’m looking forward to trying it out.

abhishek-oai reviewed (commented):
Addressed all comments

abhishek-oai commented:
Fix for CI failure (seems like a flake) - https://github.com/openai/codex/pull/3844

bolinfest commented on codex-rs/protocol/src/protocol.rs:
```diff
@@ -1298,4 +1312,19 @@ mod tests {
         assert_eq!(deserialized, event);
     }
+
+    #[test]
```
I think we can do without this test: it doesn't provide much in the way of additional code correctness.

abhishek-oai replied:
Done

bolinfest commented on codex-rs/protocol/src/protocol.rs:
```diff
@@ -1032,6 +1042,10 @@ pub struct ExecCommandBeginEvent {
+    /// True when this exec was initiated directly by the user (e.g. bang command),
+    /// not by the agent/model. Defaults to false for backwards compatibility.
+    #[serde(default)]
+    pub user_initiated_shell_command: bool,
```
How about: `is_user_shell_command` to mirror the name of the `Op`?

bolinfest commented on codex-rs/tui/src/chatwidget.rs:
```diff
@@ -97,6 +97,7 @@ use codex_protocol::mcp_protocol::ConversationId;
 struct RunningCommand {
     command: Vec<String>,
     parsed_cmd: Vec<ParsedCommand>,
+    user_initiated_shell_command: bool,
```
Rename this as well?

bolinfest commented on codex-rs/tui/src/chatwidget.rs:
```diff
@@ -999,6 +1005,15 @@ impl ChatWidget {
+        // Special-case: "!cmd" executes a local shell command instead of sending to the model.
+        if let Some(stripped) = text.strip_prefix('!') {
+            let cmd = stripped.trim().to_string();
+            if !cmd.is_empty() {
```
So if the user types `!` by itself, we'll send the message `"!"` to the agent, correct?

I guess this is fine, though we could take this opportunity to tell the user about the `!` feature instead?

bolinfest commented on codex-rs/tui/src/history_cell.rs:
```diff
@@ -466,24 +467,34 @@ impl ExecCell {
-                    body_lines.extend(wrapped.into_iter().map(|l| Line::from(l.to_string().dim())));
+
+        if let Some(output) = call.output.as_ref() {
+            let is_shell_exec_cell = self.calls.iter().any(|c| c.user_initiated_shell_command);
```
Do we believe we could have a mix of exec calls in flight, only some of which are user initiated shell commands?

bolinfest commented on codex-rs/exec/src/event_processor_with_human_output.rs:
```diff
@@ -278,6 +278,7 @@ impl EventProcessor for EventProcessorWithHumanOutput {
                 command,
                 cwd,
                 parsed_cmd: _,
+                ..
```
let's just add the extra field and assign to `_` like `parsed_cmd` so the reader immediately knows what is being ignored.

That or we delete `parsed_cmd: _,`. It's weird to use both `_` and `..`.

bolinfest commented on codex-rs/core/src/codex.rs:
```diff
@@ -1516,6 +1520,14 @@ async fn submission_loop(
+            Op::RunUserShellCommand { command } => {
+                spawn_user_shell_command_task(
+                    sess.clone(),
+                    Arc::clone(&turn_context),
```
This appears to be an unnecessary `clone()`.

```suggestion
                    &turn_context,
```

bolinfest commented on codex-rs/core/src/codex.rs:
```diff
@@ -1534,6 +1546,99 @@ async fn submission_loop(
+) {
+    let spawn_sub_id = sub_id.clone();
+    let handle = {
+        let sess = Arc::clone(&sess);
```
Prefer `.clone()` to `Arc::clone()`.

bolinfest commented on codex-rs/core/src/codex.rs:
```diff
@@ -1534,6 +1546,99 @@ async fn submission_loop(
+    sub_id: String,
+    command: String,
+) {
+    let spawn_sub_id = sub_id.clone();
```
Do this inside `let handle`, as well?

bolinfest commented on codex-rs/core/tests/suite/user_shell_cmd.rs:
```diff
@@ -0,0 +1,119 @@
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+async fn user_shell_cmd_ls_and_cat_in_temp_dir() {
+    unsafe {
+        std::env::set_var("CODEX_HERMETIC_TEST_SHELL", "1");
+    }
```
Please find another way to do this: this is `unsafe` for a reason.

bolinfest commented on codex-rs/core/tests/suite/user_shell_cmd.rs:
```diff
@@ -0,0 +1,119 @@
+use std::path::PathBuf;
+use tempfile::TempDir;
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
```
Please omit `worker_threads` unless you have a reason to believe you need them.

Empirically, tests that only pass with more than one worker thread imply bugs in our code, so I would prefer to try to catch them.

bolinfest commented on codex-rs/core/tests/suite/user_shell_cmd.rs:
```diff
@@ -0,0 +1,119 @@
+#![cfg(unix)]
```
Given these tests are limited to unix for now (which I would prefer not to do), do we really need `CODEX_HERMETIC_TEST_SHELL`?

bolinfest commented on codex-rs/core/tests/suite/user_shell_cmd.rs:
```diff
@@ -0,0 +1,119 @@
+    // 1) ls should list the file
+    codex
+        .submit(Op::RunUserShellCommand {
+            command: "ls".to_string(),
```
I often use `python3 -c` as an example for cross-platform shell calls.

abhishek-oai replied:
Done

bolinfest reviewed (dismissed):
This is generally good, but there are some things I'd like you to look at!

rtl-ai commented:
Any update on the merge timeline for this PR?

jif-oai commented on codex-rs/core/src/codex.rs:
```diff
@@ -1586,6 +1619,103 @@ async fn submission_loop(
     debug!("Agent loop exited");
 }
 
+#[derive(Clone)]
```
If this is a task, can you move it into the task module?

abhishek-oai reviewed (commented):
Pushing old comments that I had replied to

jif-oai commented on codex-rs/core/src/tasks/user_shell.rs:
```diff
@@ -0,0 +1,108 @@
+    }
+}
+
+async fn run_user_shell_command(
```
You can inline it as I guess you won't unit test it

jif-oai commented on codex-rs/core/src/tasks/user_shell.rs:
```diff
@@ -0,0 +1,108 @@
+    });
+    session.send_event(turn_context.as_ref(), event).await;
+
+    let shell_invocation = session
```
Why do you need this? The use should just answer a valid command for his shell. This should not be our responsibility to format it

jif-oai commented on codex-rs/core/src/shell.rs:
```diff
@@ -29,6 +29,96 @@ pub enum Shell {
 }
```
IMO we should revert all this file (per my previous comments)

jif-oai reviewed (approved):
After my comments, lgtm

abhishek-oai committed:
Run user shell comment

abhishek-oai committed:
spawn_user_shell_command task

abhishek-oai commented on codex-rs/core/src/codex.rs:
```diff
@@ -1228,6 +1229,19 @@ impl Session {
         &self.services.user_shell
     }
 
+    pub async fn spawn_user_shell_command(
```
Done

abhishek-oai reviewed (commented):
Fixed 2 comments

abhishek-oai committed:
core: stop formatting user shell commands; use shlex split. Remove format helpers in shell.rs and drop related test per review.

abhishek-oai committed:
core: run user-entered scripts via default shell (zsh/bash with -lc, PowerShell with -Command); fallback to shlex split only when shell is unknown.

abhishek-oai committed:
Rename variable

abhishek-oai committed:
Fmt

abhishek-oai committed:
Remove unneeded

abhishek-oai commented on codex-rs/core/src/tools/handlers/shell.rs:
```diff
@@ -108,6 +110,7 @@ impl ShellHandler {
         turn: Arc<TurnContext>,
         tracker: crate::tools::context::SharedTurnDiffTracker,
         call_id: String,
+        is_user_shell_command: bool,
```
Done

abhishek-oai committed:
Add comment for TODO
